src/contrib/mlir/lowerer.cpp

```diff
@@ -398,8 +418,130 @@ namespace
         rewriter.replaceOp(op, {result});
     }
 
+    REWRITER(NGArgMinRedOp)
+    {
+        lowerIndexReduction<mlir::NGArgMinRedOp>(op, operands, rewriter, m_pass, true);
+    }
+
+    REWRITER(NGArgMaxRedOp)
+    {
+        lowerIndexReduction<mlir::NGArgMaxRedOp>(op, operands, rewriter, m_pass, false);
+    }
+
     REWRITER(NGReturnOp) { rewriter.replaceOpWithNewOp<ReturnOp>(op); }
 #undef REWRITER
+
+    template <typename T>
+    void lowerIndexReduction(Operation* op,
+                             ArrayRef<Value*> operands,
+                             PatternRewriter& rewriter,
+                             DialectLoweringPass& m_pass,
+                             bool isMin)
+    {
+        T argmin = cast<T>(op);
+        auto loc = argmin.getLoc();
+        auto axesAttr = argmin.axes();
+
+        NGRAPH_CHECK(axesAttr.size() == 1, "Index Reduction op should have one reduction axis");
+        Attribute axisAttr = *axesAttr.begin();
+        unsigned axis = axisAttr.dyn_cast<IntegerAttr>().getInt();
+
+        NGRAPH_CHECK(operands.size() == 1 && operands[0] != nullptr,
+                     "Expected one non-null operand in Index Reduction op");
+
+        // Retrieve/generate Values for operands and result.
+        ScopedContext scope(rewriter, loc);
+        Value* arg = operands[0];
+        auto arg_type = arg->getType().cast<MemRefType>();
+
+        Value* finalResult = m_pass.buildOutputDefs(op, rewriter)[0];
+        Type type = argmin.getResult()->getType();
+        NGTensorType resultTy = type.cast<NGTensorType>();
+        // MLIR doesn't support Index to/from Integer type-conversion
+        // We have to store our result in an IndexType tensor and call-back to a type-conversion routine in nGraph
+        // TODO: Fix this once MLIR provides explicit cast operations.
```

Comment:
Can we put in this workaround only when sizeof(intptr_t) != resultTy.getElementType()...getWidth()?

Reply:
Gonna be ugly to avoid copying for argmin/max if IndexType matches IntType width. The problem is that func signature is built to say i64 is returned. Re-typing the output to IndexType will cause a mismatch with the ReturnOp. The MLIR canonicalizer fixes that mismatch silently, but that brings us back to square one. Ideally, I would have liked to just move the memref to the i64 result, but I don't think MLIR allows memref copy by pointer to avoid aliasing.
I am giving up on that for now (or could do a memcpy in the call-back instead), since upcoming casting Op should fix all this.